pkg/extension/delete_interceptor.go

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package extension
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Peripli/service-manager/pkg/types"
+
+	"github.com/Peripli/service-manager/pkg/query"
+
+	"github.com/Peripli/service-manager/storage"
+)
+
+type namedDeleteAPIFunc struct {
+	Name string
+	Func func(InterceptDeleteOnAPI) InterceptDeleteOnAPI
+}
+
+type namedDeleteTxFunc struct {
+	Name string
+	Func func(InterceptDeleteOnTx) InterceptDeleteOnTx
+}
+
+type deleteHookOnAPIHandler struct {
+	DeleteHookOnAPIFuncs []*namedDeleteAPIFunc
+	DeleteHookOnTxFuncs  []*namedDeleteTxFunc
+}
+
+func (c *deleteHookOnAPIHandler) OnAPIDelete(f InterceptDeleteOnAPI) InterceptDeleteOnAPI {
+	for i := range c.DeleteHookOnAPIFuncs {
+		f = c.DeleteHookOnAPIFuncs[len(c.DeleteHookOnAPIFuncs)-1-i].Func(f)
+	}
+	return f
+}
+
+func (c *deleteHookOnAPIHandler) OnTxDelete(f InterceptDeleteOnTx) InterceptDeleteOnTx {
+	for i := range c.DeleteHookOnTxFuncs {
+		f = c.DeleteHookOnTxFuncs[len(c.DeleteHookOnTxFuncs)-1-i].Func(f)
+	}
+	return f
+}
+
+// UnionDeleteInterceptor returns a function which spawns all delete interceptors, sorts them and wraps them into one.
+func UnionDeleteInterceptor(providers []DeleteInterceptorProvider) func() DeleteInterceptor {
+	return func() DeleteInterceptor {
+		c := &deleteHookOnAPIHandler{}
+		c.DeleteHookOnAPIFuncs = make([]*namedDeleteAPIFunc, 0, len(providers))
+		c.DeleteHookOnTxFuncs = make([]*namedDeleteTxFunc, 0, len(providers))
+
+		for _, p := range providers {
+			hook := p.Provide()
+			positionAPIType := PositionNone
+			positionTxType := PositionNone
+			nameAPI := ""
+			nameTx := ""
+
+			if orderedProvider, isOrdered := p.(Ordered); isOrdered {
+				positionAPIType, nameAPI = orderedProvider.PositionAPI()
+				positionTxType, nameTx = orderedProvider.PositionTx()
+			}
+
+			c.insertAPIFunc(positionAPIType, nameAPI, &namedDeleteAPIFunc{
```

Comment:
I would peffer if we can do without the temporary wrapper types and pass here just  func(InterceptDeleteOnAPI) InterceptDeleteOnAPI and name separately (applies for all operations)